clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -830,96 +872,110 @@ invariant_checker_t::print_results()
                     opnd_get_pc(instr_get_target(shard->prev_instr_decoded_->data)));
-                shard->branch_target_cache[prev_instr_trace_pc] = cond_branch_target;
+                shard->branch_target_cache[prev_instr_trace_pc] = branch_target;
             }
         }
     }
-    if (prev_instr_trace_pc != 0 /*first*/) {
-        // Check for all valid transitions except taken branches. We consider taken
-        // branches later so that we can provide a different message for those
-        // invariant violations.
-        const bool valid_nonbranch_flow =
-            // Filtered.
-            TESTANY(OFFLINE_FILE_TYPE_FILTERED | OFFLINE_FILE_TYPE_IFILTERED,
-                    shard->file_type_) ||
-            // Regular fall-through.
-            (prev_instr_trace_pc + shard->prev_instr_.instr.size == memref.instr.addr) ||
-            // String loop.
-            (prev_instr_trace_pc == memref.instr.addr &&
-             (memref.instr.type == TRACE_TYPE_INSTR_NO_FETCH ||
-              // Online incorrectly marks the 1st string instr across a thread
-              // switch as fetched.
-              // TODO i#4915, #4948: Eliminate non-fetched and remove the
-              // underlying instrs altogether, which would fix this for us.
-              (!knob_offline_ && shard->saw_timestamp_but_no_instr_))) ||
-            // Kernel-mediated, but we can't tell if we had a thread swap.
-            (shard->prev_xfer_marker_.instr.tid != 0 &&
-             (shard->prev_xfer_marker_.marker.marker_type ==
-                  TRACE_MARKER_TYPE_KERNEL_EVENT ||
-              shard->prev_xfer_marker_.marker.marker_type ==
-                  TRACE_MARKER_TYPE_KERNEL_XFER ||
-              shard->prev_xfer_marker_.marker.marker_type ==
-                  TRACE_MARKER_TYPE_RSEQ_ABORT)) ||
-            // We expect a gap on a window transition.
-            shard->window_transition_ ||
-            shard->prev_instr_.instr.type == TRACE_TYPE_INSTR_SYSENTER;
-
-        if (!valid_nonbranch_flow) {
-            // Check if the type is a branch instruction and there is a branch target
-            // mismatch.
-            if (type_is_instr_branch(shard->prev_instr_.instr.type)) {
-                const bool valid_branch_flow =
-                    // Indirect branches we cannot check.
-                    !type_is_instr_direct_branch(shard->prev_instr_.instr.type) ||
-                    // Conditional fall-through hits the regular case above.
-                    !have_cond_branch_target || memref.instr.addr == cond_branch_target;
-
-                if (!valid_branch_flow) {
-                    error_msg = "Direct branch does not go to the correct target";
+    // Check for all valid transitions except taken branches. We consider taken
+    // branches later so that we can provide a different message for those
+    // invariant violations.
+    bool fall_through_ok = !type_is_instr_branch(prev_instr.instr.type) ||
```

Comment:
Why the "ok"? fall_through_ok makes it sound like we have made sure that it has fallen through as expected. But this bool is just denoting whether we should expect a fall through.

Reply:
But your "we should expect a fall through" is a superset of "a fall through would be ok", right?  How about "fall_through_allowed"?  That should be clearer that it might or might not actually fall through.

Reply:
+1

Reply:
fall_through_ok could mean "a fall through would be ok" vs "the fall through was ok". So fall_through_allowed is better.